### PR TITLE
[esp32] Optimize preferences is_changed() by replacing temporary vector with unique_ptr

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

Optimize ESP32 preferences `is_changed()` method by replacing temporary `std::vector` with `std::unique_ptr` for more efficient memory comparison.

This optimization improves the performance of preference synchronization on ESP32 by eliminating unnecessary vector overhead during change detection. The `is_changed()` method is called for every preference during sync operations, so this optimization provides cumulative benefits.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Part of ongoing ESP32 memory and performance optimization efforts

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- N/A (internal implementation change, no user-facing changes)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# No configuration changes required - this is an internal optimization
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - N/A - Internal implementation change only

### Changes made:
1. Replaced temporary `NVSData` struct (containing `std::vector<uint8_t>`) with `std::unique_ptr<uint8_t[]>` in `is_changed()` method
2. Added early-exit optimization when sizes don't match (avoids allocation entirely)
3. Replaced vector comparison operator (`!=`) with direct `memcmp()` for better performance
4. Added `#include <memory>` for `std::unique_ptr` support
5. Maintains RAII guarantees with `std::unique_ptr` using `std::make_unique<uint8_t[]>()`

### Memory impact:
- **Flash**: 1,162,810 → 1,162,594 bytes (**-216 bytes saved**)
- **RAM**: 51,360 → 51,360 bytes (unchanged)

### Performance improvements:
Benchmark results show significant speedup for change detection:
- Small data (16B): **1.65x faster**
- Medium data (128B): **3.70x faster**  
- Large data (1KB): **7.44x faster**
- Very large data (4KB): **8.66x faster**
- Size mismatch (early exit): **62x faster**

### Testing performed:
- Comprehensive unit tests covering all edge cases
- Performance benchmarks demonstrating speedup across various data sizes
- Verified identical functional behavior with original implementation
- Tested with both matching and non-matching data scenarios

### Why this optimization matters:
1. **Called frequently**: The `is_changed()` method is invoked for every preference during sync operations
2. **Eliminates overhead**: Removes unnecessary `std::vector` template instantiation and management overhead
3. **Better cache locality**: Direct memory comparison with `memcmp()` is more cache-friendly
4. **Early exit optimization**: Avoids memory allocation entirely when sizes don't match
5. **Cumulative benefit**: Small improvements per call add up during bulk sync operations

### Implementation notes:
- The main `NVSData` struct still appropriately uses `std::vector` for dynamic preference caching
- Only the temporary comparison buffer was optimized
- Maintains RAII guarantees with `std::unique_ptr`
- No changes to external API or behavior

### CODE REVIEW

**Comment by bdraco** [comment]
[test_esp32_preferences.cpp.zip](https://github.com/user-attachments/files/21782800/test_esp32_preferences.cpp.zip)
[benchmark_esp32_preferences.cpp.zip](https://github.com/user-attachments/files/21782801/benchmark_esp32_preferences.cpp.zip)

File: esphome/components/esp32/preferences.cpp [hunk 1]
```diff
@@ -156,20 +157,23 @@ class ESP32Preferences : public ESPPreferences {
...
+      return true;
+    }
+    auto stored_data = std::make_unique<uint8_t[]>(actual_len);
+    err = nvs_get_blob(nvs_handle, to_save.key.c_str(), stored_data.get(), &actual_len);
     if (err != 0) {
       ESP_LOGV(TAG, "nvs_get_blob('%s') failed: %s", to_save.key.c_str(), esp_err_to_name(err));
       return true;
     }
-    return to_save.data != stored_data.data;
+    return memcmp(to_save.data.data(), stored_data.get(), actual_len) != 0;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/esp32/preferences.cpp]: That actually is better

**Review by kbx81 — APPROVED** [review]
Tested on a few devices here & no issues. Thanks @bdraco! 🍻

**Comment by bdraco** [comment]
thanks
